### ISSUE TITLE
`uv audit` Command for Security Vulnerability Scanning

### ISSUE THREAD

**Issue by 31z4** [astral-sh/uv#9189]
## Problem

Currently, there's no integrated way to audit dependencies defined in `pyproject.toml` and `uv.lock` against known security vulnerabilities in the [Python Packaging Advisory Database](https://github.com/pypa/advisory-database). This creates security risks and workflow inefficiencies:

- Dependencies across different groups cannot be audited in a single operation
- Current solutions require additional tools and complex workflows
- Cross-platform compatibility issues with existing solutions
- Risk of auditing outdated or incorrect dependency sets

## Proposed Solution

Implement a native `uv audit` command that would:

1. Read dependencies from both `pyproject.toml` and `uv.lock`
2. Check all locked dependencies (including all dependency groups) against the Python Packaging Advisory Database
3. Generate a security report highlighting:
   - Identified vulnerabilities
   - Affected versions
   - Available fixes

Example usage:
```bash
# Audit all dependencies
uv audit

# Audit specific dependency groups
uv audit --group dev,test

# Output in different formats
uv audit --format json
```

## Benefits

- **Improved Security**: Integrated security scanning as part of the uv toolchain
- **Better Performance**: Native implementation could offer significant speed improvements (assuming that querying the Python Packaging Advisory Database isn't a bottleneck)
- **Cross-Platform**: Works consistently across all supported platforms
- **Comprehensive Coverage**: 
  - Seamless support for dependency groups
  - Covers all project dependencies, not just installed packages
  - Ensures audit matches locked dependencies exactly
- **Simplified Workflow**: 
  - No additional tools required
  - Single command for complete package security audit
  - Native integration with uv's dependency management
  
### 1. Using pip-audit
- **Current Status**: 
  - Limited `pyproject.toml` support (ref: https://github.com/pypa/pip-audit/issues/766)
  - No `uv.lock` support
- **Limitations**:
  - Additional dependency requirement
  - Cross-platform inconsistencies
  - Incomplete dependency group coverage

### 2. Extending pip-audit for `uv.lock` support
- **Pros**:
  - Leverages existing tool
  - Community familiarity
- **Cons**:
  - Loses uv's performance benefits
  - Additional dependency requirement remains
  - More complex integration

### 3. Local environment scanning with pip-audit
- **Approach**: Running `pip-audit -l` against installed packages
- **Issues**:
  - Environment may not match declarations
  - Limited dependency group coverage
  - Platform-specific behavior

### 4. Requirements.txt generation and scanning with pip-audit
- **Approach**: Convert `uv.lock` to `requirements.txt` format for scanning
- **Issues**:
  - Additional synchronization overhead
  - Platform compatibility issues
  - Complex workflow

## Additional Context

- Related Issue: Security check before installation discussion (#8842)
- Implementation Interest: I'm willing to contribute to the implementation pending community feedback
- Similar Features: 
  - cargo audit (Rust)
  - npm audit (Node.js)
  - composer audit (PHP)

## Next Steps
1. Gather community feedback on the proposed approach
2. Discuss implementation details if approved
3. Define specific behavior for edge cases
4. Determine output format standards

---
💡 Please share your thoughts on this proposal, particularly regarding:
- Preferred output formats
- Specific use cases to consider

**Comment by woodruffw** [comment]
For prior art: take a look at [pip-audit](https://github.com/pypa/pip-audit), which currently interoperates with `pip` and the other PyPA tooling.

Based on our experience with that tool, some scattered thoughts:

* *Avoid security fatigue*: all developers are susceptible to security fatigue, and auditing tools *in particular* eat into the "fatigue budget" more than most other tools. We try to minimize fatigue in `pip-audit` by making it an *independent flow*, i.e. not spitting out warnings or errors on every `pip install`, since we think users are more likely to just ignore security advisories if we force them in their faces.
* *Be aware of quality issues*: the PyPA advisory database is curated, but low-quality submissions and vulnerability spam/reputation-seeking chum can still make its way in. Having a system by which end users can ignore obviously bogus or irrelevant reports is a good day 1 feature to have :slightly_smiling_face: 
* *Be aware of current fidelity gaps in Python advisory reporting*: Python package advisories currently operate at the `(name, version)` scope, but many vulnerabilities are more specific than that: they either apply only to binary builds (wheels) of packages, or onto to specific API surfaces, etc. There's a long-term plan to increase the fidelity in reports (e.g. to expose which files within a release are vulnerable, or which API surfaces are vulnerable), but it's worth noting that the current level of fidelity *will* cause false positives.

**Comment by saada** [comment]
In the meantime, this seems to work fairly well

```
uvx pip-audit -r requirements.txt --fix
```

**Comment by strayer** [comment]
I had some issues with pip-audit trying to install pip (`subprocess.CalledProcessError: Command '['/var/folders/n9/kcz03p6n2pb5xchxn1ywm4640000gn/T/tmp74_otz01/bin/python3.11', '-m', 'ensurepip', '--upgrade', '--default-pip']' died with <Signals.SIGABRT: 6>`), so this is what I'm doing for now:

`uv export --format requirements-txt > requirements.txt && uvx pip-audit -r requirements.txt --disable-pip && rm requirements.txt`

Seems to work very well too.

**Comment by aplantinga** [comment]
See also the [uv-secure](https://github.com/owenlamont/uv-secure) project, which seems to have a similar aim of providing pip-audit functionality for uv

**Comment by colin-kerkhof** [comment]
I have been running `uv run --all-groups --with pip-audit pip-audit -l`. That should make sure that every package in the `pyproject.toml` is checked against the database, no?

**Comment by ycheng517** [comment]
> I had some issues with pip-audit trying to install pip (`subprocess.CalledProcessError: Command '['/var/folders/n9/kcz03p6n2pb5xchxn1ywm4640000gn/T/tmp74_otz01/bin/python3.11', '-m', 'ensurepip', '--upgrade', '--default-pip']' died with <Signals.SIGABRT: 6>`), so this is what I'm doing for now:
> 
> `uv export --format requirements-txt > requirements.txt && uvx pip-audit -r requirements.txt --disable-pip && rm requirements.txt`
> 
> Seems to work very well too.

I get an error when I try this:

```
uv export --no-emit-project --format requirements-txt > requirements.txt && uvx pip-audit -r requirements.txt --disable-pip
Resolved 491 packages in 338ms
Traceback (most recent call last):
  File "/home/yifei/.cache/uv/archive-v0/ebnZSnxmb7JKYb1q41oSp/bin/pip-audit", line 12, in <module>
    sys.exit(audit())
             ~~~~~^^
  File "/home/yifei/.cache/uv/archive-v0/ebnZSnxmb7JKYb1q41oSp/lib/python3.13/site-packages/pip_audit/_cli.py", line 538, in audit
    for spec, vulns in auditor.audit(source):
                       ~~~~~~~~~~~~~^^^^^^^^
  File "/home/yifei/.cache/uv/archive-v0/ebnZSnxmb7JKYb1q41oSp/lib/python3.13/site-packages/pip_audit/_audit.py", line 68, in audit
    for dep, vulns in self._service.query_all(specs):
                      ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^
  File "/home/yifei/.cache/uv/archive-v0/ebnZSnxmb7JKYb1q41oSp/lib/python3.13/site-packages/pip_audit/_service/interface.py", line 155, in query_all
    for spec in specs:
                ^^^^^
  File "/home/yifei/.cache/uv/archive-v0/ebnZSnxmb7JKYb1q41oSp/lib/python3.13/site-packages/pip_audit/_dependency_source/requirement.py", line 134, in collect
    yield from self._collect_from_files(collect_files)
  File "/home/yifei/.cache/uv/archive-v0/ebnZSnxmb7JKYb1q41oSp/lib/python3.13/site-packages/pip_audit/_dependency_source/requirement.py", line 167, in _collect_from_files
    yield from self._collect_preresolved_deps(iter(reqs), require_hashes)
  File "/home/yifei/.cache/uv/archive-v0/ebnZSnxmb7JKYb1q41oSp/lib/python3.13/site-packages/pip_audit/_dependency_source/requirement.py", line 319, in _collect_preresolved_deps
    if req.marker is not None and not req.marker.evaluate():
                                      ~~~~~~~~~~~~~~~~~~~^^
  File "/home/yifei/.cache/uv/archive-v0/ebnZSnxmb7JKYb1q41oSp/lib/python3.13/site-packages/packaging/markers.py", line 319, in evaluate
    return _evaluate_markers(
        self._markers, _repair_python_full_version(current_environment)
    )
  File "/home/yifei/.cache/uv/archive-v0/ebnZSnxmb7JKYb1q41oSp/lib/python3.13/site-packages/packaging/markers.py", line 225, in _evaluate_markers
    groups[-1].append(_eval_op(lhs_value, op, rhs_value))
                      ~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/yifei/.cache/uv/archive-v0/ebnZSnxmb7JKYb1q41oSp/lib/python3.13/site-packages/packaging/markers.py", line 183, in _eval_op
    return spec.contains(lhs, prereleases=True)
           ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/yifei/.cache/uv/archive-v0/ebnZSnxmb7JKYb1q41oSp/lib/python3.13/site-packages/packaging/specifiers.py", line 552, in contains
    normalized_item = _coerce_version(item)
  File "/home/yifei/.cache/uv/archive-v0/ebnZSnxmb7JKYb1q41oSp/lib/python3.13/site-packages/packaging/specifiers.py", line 28, in _coerce_version
    version = Version(version)
  File "/home/yifei/.cache/uv/archive-v0/ebnZSnxmb7JKYb1q41oSp/lib/python3.13/site-packages/packaging/version.py", line 202, in __init__
    raise InvalidVersion(f"Invalid version: {version!r}")
packaging.version.InvalidVersion: Invalid version: '6.11.0-21-generic'
```

It looks to be caused by the presence of markers in requirements such as `platform_release >= '20.0' and sys_platform == 'darwin'`

**Comment by woodruffw** [comment]
> It looks to be caused by the presence of markers in requirements such as `platform_release >= '20.0' and sys_platform == 'darwin'`

That's not why -- the error message shows that the error is caused by an invalid package version:

```
  File "/home/yifei/.cache/uv/archive-v0/ebnZSnxmb7JKYb1q41oSp/lib/python3.13/site-packages/packaging/version.py", line 202, in __init__
    raise InvalidVersion(f"Invalid version: {version!r}")
packaging.version.InvalidVersion: Invalid version: '6.11.0-21-generic'
```

You can reproduce this directly as well:

```
uv run --with packaging python
>>> import packaging.version
>>> packaging.version.Version("6.11.0-21-generic")
packaging.version.InvalidVersion: Invalid version: '6.11.0-21-generic'
```

If you could report that upstream to `pip-audit` with a reproduction case, I'd appreciate it -- `pip-audit` *should* be catching and skipping these kinds of invalid package versions, but it's always possible there are edge cases we missed.

You can report it here: https://github.com/pypa/pip-audit/issues/new?template=bug-report.yml

**Comment by ycheng517** [comment]
> That's not why -- the error message shows that the error is caused by an invalid package version:
> 
> ```
>   File "/home/yifei/.cache/uv/archive-v0/ebnZSnxmb7JKYb1q41oSp/lib/python3.13/site-packages/packaging/version.py", line 202, in __init__
>     raise InvalidVersion(f"Invalid version: {version!r}")
> packaging.version.InvalidVersion: Invalid version: '6.11.0-21-generic'
> ```

'6.11.0-21-generic' is my Linux kernel version, this version doesn't exist at all in my `requirements.txt` file. I believe because of the `platform_version` marker, pip-audit tries to get the version of the OS kernel to compare, and fail to recognize it as a valid version. Your point about this being a `pip-audit` issue still stands though, I shall report there. On the UV side, if there's an option to not generate the markers, that will help as well.

**Comment by woodruffw** [comment]
Oh huh, that's interesting! Yeah, please file an issue, thanks -- that _might_ even be a lower-level issue than `pip-audit` itself (since assuming that `platform_version` evaluates to something like a PEP 440 version seems wrong within `packaging`).

**Comment by ycheng517** [comment]
Ah, thanks for finding that issue!

**Comment by Avasam** [comment]
I personally would like to get vulnerability warnings whenever I install (sync, add, uv pip) à la npm. But I'm fine if its opt in rather than opt out.

**Comment by owenlamont** [comment]
I should've searched for this issue earlier - I am the author of [uv-secure](https://github.com/owenlamont/uv-secure) - but I'd love for my package to become obsolete if uv did add an audit sub-command or perform automatic vulnerability scanning when syncing or locking.

In my case uv-secure hits the PyPi json API to request known vulnerabilities per package version and requesting that info in parallel / asynchronously can still take a while - I'm sure the uv devs could do that smarter and more efficiently but you may not want that option always enabled by default. Maybe a _uv audit_ command would be the way to go.

In the short-term if I can find ways to incorporate more of uv into uv-secure (uv-secure is currently pure Python and not as fast as I'd like) I'd be interested to learn ways I could do that.

**Comment by nyudenkov** [comment]
@31z4 @owenlamont @Avasam @woodruffw 

Hi! Inviting you to take a look at #15053 :)

**Comment by nyudenkov** [comment]
Unfortunately, the uv team can't review and accept my solution right now, but I published it as a standalone solution until uv has it built-in

https://github.com/nyudenkov/pysentry

**Comment by DEKHTIARJonathan** [comment]
Oh damn ! @nyudenkov that's awesome !

**Comment by thomasschafer** [comment]
@nyudenkov if you don't mind me asking, do the uv team have general reservations against anyone external from implementing this feature, or was it something specific about that PR that they were unsure about?

**Comment by nyudenkov** [comment]
@thomasschafer 
Nothing special about PR actually but for other details inviting @zanieb to answer

**Comment by zanieb** [comment]
Yeah we just don't have the bandwidth to do the design work we'd want to before adding a feature of this size yet, so we're not willing to accept any external contributions for it right now.

**Comment by whats2000** [comment]
I use a temporary workaround: emulate `uv audit` in PowerShell

Since `uv` does not currently provide a built-in `audit` command, I make it temporarily alias it to `uv-secure` in PowerShell.

This allows:

```powershell
uv audit
```

to run:

```powershell
uv tool run uv-secure uv.lock
```

<img width="922" height="232" alt="Image" src="https://github.com/user-attachments/assets/b69972c6-d409-4c2d-b805-8ab0a046ba3a" />

---

If anyone interesting here is the step for Windows:

### Step 1 — Open your PowerShell profile

```powershell
notepad $PROFILE
```

If the file does not exist:

```powershell
New-Item -Type File -Path $PROFILE -Force
notepad $PROFILE
```

### Step 2 — Add the following code

Append this to the file:

```powershell
# Save the uv.exe path
$Global:RealUv = (Get-Command uv.exe).Source

function uv {
    param(
        [Parameter(ValueFromRemainingArguments = $true)]
        [string[]]$Args
    )

    if ($Args.Count -gt 0 -and $Args[0] -eq "audit") {
        Write-Host "Running uv-secure audit..." -ForegroundColor Cyan
        & $Global:RealUv tool run uv-secure uv.lock
        return
    }

    & $Global:RealUv @Args
}
```

Save and close the file.

### Step 3 — Restart PowerShell

Close all PowerShell windows and open a new one.

(This is required for the profile changes to take effect.)

---

### Notes

* This is a local PowerShell-only workaround.
* It does not modify `uv` itself.
* Intended as a temporary solution until native audit support exists.
* For CI / cross-platform setups, use `uv-secure` or `pip-audit` directly.
* Reference: https://github.com/owenlamont/uv-secure

**Comment by owenlamont** [comment]
I'm flattered you use uv-secure @whats2000 but I don't think trying alias tricks to make it look like it's part of uv will go down well (I personally run it as a pre-commit tool with prek on commit automatically).

I can't  speak for astral devs but my understanding is their new pyx platform will do this sort of package vulnerability scanning server side (which will definitely be much more efficient) so if you use uv with pyx I imagine you get this out of the box and it would be redundant for uv to locally scan for vulnerabilities so my guess is this issue will be closed as redundant with what pyx does.

**Comment by Avasam** [comment]
> I use a temporary workaround: emulate uv audit in PowerShell

That's... quite an over complication over just running `uvx uv-secure` or `uvx pysentry-rs`...

That post also has a few hallmarks of LLM-generation.

**Comment by whats2000** [comment]
> That post also has a few hallmarks of LLM-generation.

Yes, I did ask for a step guide to do that. I know it seems like redundant, but for other package manager I used is mostly in that format (I common work with `npm` and `pnpm`). 

What I hope is to make it solve dependence and suggest update just like how it work in `npm audit fix`. 

But, after the reaction by you guys, I think maybe yep... it just redundant to have a command like that.

**Comment by woodruffw** [comment]
> if you use uv with pyx I imagine you get this out of the box and it would be redundant for uv to locally scan for vulnerabilities so my guess is this issue will be closed as redundant with what pyx does.

JFYI we *are* working on a design for this, and should have some more to share soon. It's true that pyx has similar functionality, but a user's lockfile is always the "ground truth" for an application's dependencies and hence local vulnerability scanning always has value (including outside of users of pyx, of course).

**Comment by woodruffw** [comment]
Hey folks, you can see our current roadmap for this in #18506. TL;DR is that we've completed all of our own MVP priorities here, and you can begin experimenting with `uv audit` today.

Please keep in mind that we'll keep it in preview for a bit, to allow us to shake out any bugs and design decisions that need iteration!

**Comment by riley-lunar** [comment]
Mentioning this here for CLI design coherence, but would it be worthwhile to make a `uv deps` command with `uv deps audit` and `uv deps check-license` under it? I don't really care about what they're called but both cve check and license check don't fall under the following:
- `uv pip`: not a pip operation
- `uv tree`: not a tree representation necessarily

And both of them deal with some extra checks that uv handles on the dependency list itself. This would also lend itself nicely to config sections in the `pyproject.toml`.

https://github.com/astral-sh/uv/pull/10292

**Comment by LecrisUT** [comment]
Interesting feature here. Does anyone now if there is work on integrating this with github actions workflows. We have an issue with python packages that need system devel packages and dependabot always reports an error in those cases ([example](https://github.com/teemtee/tmt/security/dependabot/15)), and would be nice to have an alternative that actually works well.

**Comment by quantumpacket** [comment]
@owenlamont I see you deprecated your `uv-secure` tool. Maybe I'm overlooking things, but I don't see any options for `uv audit` that replicate the following options that your tool provided.


```
--forbid-archived
--forbid-deprecated
--forbid-quarantined
--forbid-yanked
```

Does anyone know if these kind of restrictions are baked into `uv audit` and are just not configurable? Or is it something that perhaps should be added in a future release?

**Comment by owenlamont** [comment]
> @owenlamont I see you deprecated your `uv-secure` tool. Maybe I'm overlooking things, but I don't see any options for `uv audit` that replicate the following options that your tool provided.
> 
> 
> ```
> --forbid-archived
> --forbid-deprecated
> --forbid-quarantined
> --forbid-yanked
> ```
> 
> Does anyone know if these kind of restrictions are baked into `uv audit` and are just not configurable? Or is it something that perhaps should be added in a future release?

You should check the UV audit roadmap issue https://github.com/astral-sh/uv/issues/18506

That's probably the best place to ask.

**Comment by Owen-OptiGrid** [comment]
Think https://github.com/astral-sh/uv/pull/19128 is part way towards addressing that @quantumpacket - if you keep checking the roadmap issue I mentioned above sounds like that is in the works.
